### PR TITLE
[preview] app proto v14

### PR DESCRIPTION
**Pull request by inliniac**
Protocol detection and matching update. Fixes bug https://redmine.openinfosecfoundation.org/issues/1690

Fixes negated app-layer-protocol matching and various other issues with the keyword. Improves tests. Implements prefilter.

Couple of things that changed:
- log protocol per direction in addition to the 'app_proto': app_proto_ts and app_proto_tc
- explicitly log 'failed' if detection failed for a direction or completely
- add a 'failed' protocol counter similar to the app-layer counters

Prscript:
- PR inliniac-pcap: https://buildbot.openinfosecfoundation.org/builders/inliniac-pcap/builds/11
- PR inliniac: https://buildbot.openinfosecfoundation.org/builders/inliniac/builds/519

### CODE REVIEW

File: src/output-json-flow.c [hunk 1]
```diff
@@ -187,7 +187,12 @@ static void JsonFlowLogJSON(JsonFlowLogThread *aft, json_t *js, Flow *f)
         return;
     }
 
-    json_object_set_new(js, "app_proto", json_string(AppProtoToString(f->alproto)));
+    json_object_set_new(js, "app_proto",
+            json_string(AppProtoToString(f->alproto)));
+    json_object_set_new(js, "app_proto_ts",
+            json_string(AppProtoToString(f->alproto_ts)));
+    json_object_set_new(js, "app_proto_tc",
+            json_string(AppProtoToString(f->alproto_tc)));
```
Comment (jasonish) [inline src/output-json-flow.c]: I'm wondering if we want this extra information if they are all equal to alproto?

**Comment by inliniac** [comment]
On 14-10-16 19:38, Jason Ish wrote:
> *@jasonish* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/output-json-flow.c
> <https://github.com/inliniac/suricata/pull/2359#pullrequestreview-4312179>:
> 
>> @@ -187,7 +187,12 @@ static void JsonFlowLogJSON(JsonFlowLogThread *aft, json_t *js, Flow *f)
>          return;
>      }
>  
> -    json_object_set_new(js, "app_proto", json_string(AppProtoToString(f->alproto)));
> +    json_object_set_new(js, "app_proto",
> +            json_string(AppProtoToString(f->alproto)));
> +    json_object_set_new(js, "app_proto_ts",
> +            json_string(AppProtoToString(f->alproto_ts)));
> +    json_object_set_new(js, "app_proto_tc",
> +            json_string(AppProtoToString(f->alproto_tc)));
> 
> I'm wondering if we want this extra information if they are all equal to
> alproto?

They can be different, e.g. when we have an http request to an ssh
server. But not sure if it's valuable to unconditionally have it either.

It maybe also be interesting when we detected only one side, and the
other side 'failed'.

**Comment by inliniac** [comment]
On 14-10-16 19:25, Jason Ish wrote:
> *@jasonish* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/detect.c
> <https://github.com/inliniac/suricata/pull/2359#pullrequestreview-4309916>:
> 
>> @@ -1998,6 +1998,103 @@ int SignatureIsIPOnly(DetectEngineCtx *de_ctx, Signature *s)
>      return 1;
>  }
>  
> +/** \internal
> + *  \brief Test is a initialized signature is PD only
> 
> Whats PD? In SignatureIsDEOnly DE is spelled out.

Protocol Detection. I'll make it explicit.

**Comment by jasonish** [comment]
> They can be different, e.g. when we have an http request to an ssh
> server. But not sure if it's valuable to unconditionally have it either.
> 
> It maybe also be interesting when we detected only one side, and the
> other side 'failed'.

Yeah, if they are different I want to know.  If they are the same then its just redundant information in the log.  Hard to say...

If ts and tc are different, what should "app_proto" by?

**Comment by inliniac** [comment]
On 14-10-16 19:48, Jason Ish wrote:
>     They can be different, e.g. when we have an http request to an ssh
>     server. But not sure if it's valuable to unconditionally have it either.
> 
>     It maybe also be interesting when we detected only one side, and the
>     other side 'failed'.
> 
> Yeah, if they are different I want to know. If they are the same then
> its just redundant information in the log. Hard to say...
> 
> If ts and tc are different, what should "app_proto" by?

The protocol detection code sets it. The logic is largely:

1. if first direction lead to the applayer parser already running,
alstate setup, then we pick that as the final.

2. if parser not yet invoked, we trust the server, not the client

3. in either case the event
app-layer-event:applayer_mismatch_protocol_both_directions; is set.

**Comment by inliniac** [comment]
Replaced by #2360

**Pull request closed without merging**